### PR TITLE
Logical Operators

### PR DESCRIPTION
**Pull request by wenxuanqiu**
Bring over logical and physical operators from Peloton. These objects are direct inputs to the optimizer. These objects are constructed by the binder from parser expressions.

Part I How to bring over individual operators
The original operators are in:
https://github.com/cmu-db/peloton/blob/master/src/include/optimizer/operators.h
https://github.com/cmu-db/peloton/blob/master/src/optimizer/operators.cpp

The physical operators have been ported over already. The ones that remain are some logical operators. They are:
LogicalFilter
LogicalProjection
LogicalDependentJoin
LogicalMarkJoin
LogicalSingleJoin
LogicalInnerJoin
LogicalLeftJoin
LogicalRightJoin
LogicalOuterJoin
LogicalSemiJoin 
LogicalAggregateAndGroupBy
LogicalInsertSelect
LogicalInsert
LogicalMarkJoin
LogicalDependentJoin
LogicalSIngleJoin
LogicalInnerJoin
LogicalLeftJoin
LogicalRightJoin
LogicalOuterJoin
LogicalSemiJoin
LogicalAggregateAndGroupBy
LogicalInsert
LogicalInsertSelect
LogicalDelete
LogicalUpdate
LogicalLimit
LogicalDistinct
LogicalExportExternalFile

Clone my repository `git clone git@github.com:wenxuanqiu/terrier.git`
Then modify only `src/include/operators.h` and `src/include/operators.cpp`. Find the operator you want to work on, then do the following:

1. Copy the corresponding Peloton operator over. Currently, all the operators that Peloton had are in `src/include/operator.h`, but not `src/include/operator.cpp`.
2. Make it compile by changing a number of things:
     - expression::AbstractExpression -> parser::AbstractionExpression
     - hash_t -> common::hash_t
     - string reference to value std::string & -> std::string then use std::move in constructor definition (operator.cpp)
     - replace std::shared_ptr<catalog::TableCatalogEntry> target_table with catalog::db_oid_t database_oid, catalog::namespace_oid_t namespace_oid, catalog::table_oid_t table_oid
     - vector reference to lvalue: std::vector<std::shared_ptr<expression::AbstractExpression>> &columns -> std::vector<std::shared_ptr<parser::AbstractExpression>> &&columns then use std::move in constructor definition (operator.cpp)
     - other case by case changes
3. Make the class members private and append '_' suffix to indicate that
4. Add oxygen comments
5. Implement the constructor, Hash, and operator in operators.cpp. Some examples are in the file.
6. Fix remaining clang-tidy problems that show up due to `check clang-tidy`
7. Commit the change

Part II
Step 2:
We need DDL statements such as CREATE and DROP to have operators as well. So we need to create them, some of them might have logical and physical versions. For example, LogicalCreateIndex vs. CreateBwTreeIndex & CreateHashIndex & CreateAVLIndex. These operators must have the same interface as the existing operators.

### CODE REVIEW

**Comment by apavlo** [comment]
@wenxuanqiu One additional thing that we should discuss on Tuesday is that we need to include the list of index OIDs that the INSERT/UPDATE/DELETE queries will want to modify. The `SQLTable` will not know anything about indexes.

File: src/include/optimizer/operator_node.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+#include <string>
+#include <vector>
+#include "common/hash_util.h"
+#include "optimizer/property_set.h"
+
+namespace terrier::optimizer {
+
+enum class OpType {
+  Undefined = 0,
+  DummyScan, /* Dummy Op for SELECT without FROM*/
```
Comment (apavlo) [inline src/include/optimizer/operator_node.h]: Maybe we should call this `TableFreeScan`?

File: src/include/parser/expression/abstract_expression.h [hunk 2]
```diff
@@ -67,7 +67,7 @@ class AbstractExpression {
   /**
    * Logical inequality check.
    * @param rhs other
-   * @return true if the two expressions are not logically equal
```
Comment (apavlo) [inline src/include/parser/expression/abstract_expression.h]: Why did you switch this?

**Comment by lmwnshn** [comment]
Is this ready for review? @wenxuanqiu

**Comment by apavlo** [comment]
This PR is now in branch [logical-ops](https://github.com/cmu-db/terrier/tree/logical-ops).

Documentation is available on [wiki](https://github.com/cmu-db/terrier/wiki/Porting-Logical-Operators).

**Pull request closed without merging**